Source/Csla.AspNetCore/ConfigurationExtensions.cs

```diff
@@ -36,6 +40,9 @@ public static CslaOptions AddAspNetCore(this CslaOptions config, Action<AspNetCo
     {
       var localOptions = new AspNetCoreConfigurationOptions();
       options?.Invoke(localOptions);
+#if NET5_0_OR_GREATER
+      config.Services.AddScoped(typeof(CircuitHandler), typeof(ActiveCircuitHandler));
```

Comment:
@TheCakeMonster the circuit handler is now being registered.

@swegele I read a few things about this. It does appear that if someone registered a singleton CircuitHandler that it would brake CSLA. It appears to be a limitation of .NET DI, so I don't know how to fix the issue.

Were someone to choose to implement a singleton handler, they'd have to figure out a way to implement a singleton solution to the CSLA requirement as well - and maybe that's something we'll solve if/when someone actually does such a thing :)

Reply:
Well before bed, I hopped on PC and ran a test.  I made an exact copy of my "scoped" circuit handler and named it with 2.  Then I registered it and put a breakpoint in OnCircuitOpenedAsync in both.

When debugging **they both get called** (interesting)...but it seems that DI will only inject the **LAST ONE** that is registered in Program.cs.  So when you have this in your page...`@inject CircuitHandler _circuitHandler;` you get the last one.  

So when I have them like this - then `UserScopedCircuitHandler` ends up `null` when I try to cast it.
```cs
builder.Services.AddScoped<CircuitHandler, UserScopedCircuitHandler>();
builder.Services.AddScoped<CircuitHandler, UserScopedCircuitHandler2>();
```


But when I reverse it, then I do get the cast just fine.
```cs
builder.Services.AddScoped<CircuitHandler, UserScopedCircuitHandler2>();
builder.Services.AddScoped<CircuitHandler, UserScopedCircuitHandler>();
```

So I'm not sure how you would ever end up injecting yours in the framework if mine is registered after it.

Reply:
So there can be only one circuit handler registered period - like most services. This, then, is fine with me. If someone needs a custom circuit handler, they'll have to incorporate the CSLA functionality into their implementation as well.

Reply:
Oh wait. I just read what you said more closely.

I think @TheCakeMonster solved this already - if you want to have more than one injected you need to request an array of the type rather than just the type.

```c#
// constructor
public MyType(CircuitHandler[] handlerList)
{
}
```

Reply:
One additional thought, since all instances get called but only one gets served up via DI...you could have your circuit handler inject to another service (e.g. `CircuitWatcher` ).  That could be the service that you use internally which could be the messenger and safe to use.  
Then us user's wouldn't have to implement anything in our circuit handlers.

Reply:
When I try
`public MyType(CircuitHandler[] handlerList)`

I get this error:
`Unable to resolve service for type 'Microsoft.AspNetCore.Components.Server.Circuits.CircuitHandler[]'`

Reply:
Viola!  Just tried my idea and I think it will work for you.  

Given this:
```cs
builder.Services.AddScoped<CircuitHandler, CslaScopedCH>();
builder.Services.AddScoped<CslaCHMessenger>();
builder.Services.AddScoped<CircuitHandler, MyScopedCH>();
```

and this additional service:
```cs
public class CslaCHMessenger
{
    public string CircuitIdFromCslaCHMessenger { get; set; }

    public CslaCHMessenger()
    {
    }
}
```

Anywhere in my app where I call `@inject CircuitHandler _circuitHandler;` I get a `MyScopedCH`.  Good.

But your `CslaScopedCH` still gets called just when the `OnCircuitOpenedAsync` event fires.

So in `CslaScopedCH` constructor you can inject `CslaCHMessenger`.  Just as a test I set the `CircuitIdFromCslaCHMessenger`  property of it.  In the Counter.razor page I inject `CslaCHMessenger` and I see it's value is set.

You could do this and have the messenger class obtain and carry whatever values out of your circuit handler that you need to test.  And you can always count on it working.

HTH

Reply:
Loads of questions and limited time, so I will try to summarise.

1. Yes, there can be multiple circuit handlers registered. The events on all of them will be raised by the underlying infrastructure.
2. If you ask for `IEnumerable<CircuitHandler>` to be injected, you get an enumerable of all of them. We know this works, as this is what we do with Interceptors (and is proven to work by the results of the Csla.Tests project.)
3. It looks like `CircuitHost` gets passed an array of `CircuitHandler` and my guess was that this was coming from DI. However, I don't recognise the syntax being used, so I can't be sure that will definitely work. Maybe CircuitHost isn't created by DI; I remember researching it and writing it all down on this repo somewhere, but I don't know exactly where TBH.
3. If you ask for a type of CircuitHandler to be injected, you get the last one registered, whatever that was. That makes it very hit and miss what you will get, so I wouldn't try this.
4. If you ask for your own derived type by name, I'm not sure what happens. At a guess it would need to be registered twice, and it might not do what you expect.

I know that my demo code worked, so it's worth checking over that again. I used a separate class to carry the data from one place to another; at a glance this also seems to be what @swegele is suggesting. Somehow that feels more correct anyway, to break the dependencies apart, but I'm not sure that I can explain _exactly_ why in the limited amount of time I have at the moment.

Reply:
It think @TheCakeMonster's solution with the messenger class is correct. I was trying to simplify things by removing it, but clearly it is necessary (or is the simplest solution).

I'll make that change to the codebase and see what we all think.